### ISSUE TITLE
Regression: unexpected com.fasterxml.jackson.databind.exc.MismatchedInputException

### ISSUE THREAD

**Issue by quaff** [FasterXML/jackson-databind#2961]
**Describe the bug**
My test works fine with 2.11.x, but failed with 2.12.0

com.fasterxml.jackson.databind.exc.MismatchedInputException: Cannot construct instance of `test.DeserializationTest$ClientData` (although at least one Creator exists): cannot deserialize from Object value (no delegate- or property-based Creator)

**Version information**
Which Jackson version(s) was this for?
`2.12.0-rc1` `2.12.0-rc2` `2.12.0`

**To Reproduce**
[jackson-test.zip](https://github.com/FasterXML/jackson-databind/files/5621164/jackson-test.zip)

```
package test;

import static org.junit.jupiter.api.Assertions.assertEquals;

import java.io.IOException;
import java.util.Objects;

import org.junit.jupiter.api.Test;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.module.paramnames.ParameterNamesModule;

public class DeserializationTest
{
	public static final ObjectMapper JSON_OBJECTMAPPER = new ObjectMapper().registerModule(new ParameterNamesModule());

	@Test
	public void test() throws Exception {
		String clientDataJson = "{\"challenge\":\"who?\",\"answer\":\"me\"}";
		String clientJson = "{\"data\":\"" + clientDataJson.replaceAll("\"", "\\\\\"") + "\"}";
		ClientData clientData = JSON_OBJECTMAPPER.readValue(clientDataJson, ClientData.class);
		Client client = JSON_OBJECTMAPPER.readValue(clientJson, Client.class);
		assertEquals(clientData, client.getData());
		assertEquals("who?", client.getData().getChallenge());
		assertEquals("me", client.getData().getAnswer());
	}

	public static class Client {
		private ClientData data;

		public ClientData getData() {
			return data;
		}

		public void setData(ClientData data) {
			this.data = data;
		}
	}

	public static class ClientData {
		private final String challenge;
		private final String answer;

		public ClientData(String challenge, String answer) {
			this.challenge = challenge;
			this.answer = answer;
		}

		public String getChallenge() {
			return challenge;
		}

		public String getAnswer() {
			return answer;
		}

		@JsonCreator
		public static ClientData valueOf(String input) throws IOException {
			// decryption input happens here
			return JSON_OBJECTMAPPER.readValue(input, ClientData.class);
		}
	}
}
```

**Comment by kupci** [comment]
Possibly same as #2962

**Comment by cowtowncoder** [comment]
This looks bit different from #2962, although might be related.

On definitions: I think the problem is that class `ClientData` has two Creator methods (one constructor, one factory method), but only one is annotated -- they both really should be. So using:

```java
    @JsonCreator
    public ClientData(String challenge, String answer) { }
```

makes tests pass (as long as build is configured to include parameter name information, I had some issues with that).

I will leave this open to think about why 2.11 would find the unannotated creator: discovery should be fine if there is just one unambiguous creator, but here we have two. One of them is factory method so perhaps that is the reason... but I am not 100% sure that would be intended behavior.

So I would recommended adding the other `@JsonCreator` at this point, even if this is something to resolve.

**Comment by quaff** [comment]
Yes, adding `@JsonCreator` on constructor fixed my real project.

**Comment by cowtowncoder** [comment]
Quick note: I now do think that fix for #2962 probably solves your problem too: the issue was that in 2.12, logic for auto-detection accidentally changed to bit more restrictive (see #2962 for more details if interested).
